File: swift/ITSClient/Sources/ITSMobility/Models/DENM/AlacarteContainer.swift
/*
 * Software Name : ITSClient
 * SPDX-FileCopyrightText: Copyright (c) Orange SA
 * SPDX-License-Identifier: MIT
 *
 * This software is distributed under the MIT license,
 * see the "LICENSE.txt" file for more details or https://opensource.org/license/MIT/
 *
 * Software description: Swift ITS client.
 */

import Foundation

/// The "à la carte" container.
public struct AlacarteContainer: Codable, Sendable {
    /// The lane position.
    public let lanePosition: Int8?
    /// The positioning solution.
    public let positioningSolution: PositioningSolution?

    private static let minLanePosition: Int8 = -1
    private static let maxLanePosition: Int8 = 14

    enum CodingKeys: String, CodingKey {
        case lanePosition = "lane_position"
        case positioningSolution = "positioning_solution"
    }

    init(lanePosition: Int8?, positioningSolution: PositioningSolution?) {
        self.lanePosition = lanePosition.map {
            clip($0, Self.minLanePosition, Self.maxLanePosition)
        }
        self.positioningSolution = positioningSolution
    }
}

/// The positioning solution.
public enum PositioningSolution: Int, Codable, Sendable {
    case noPositioningSolution = 0
    case sGNSS = 1
    case dGNSS = 2
    case sGNSSplusDR = 3
    case dGNSSplusDR = 4
    case dR = 5
}

File: java/iot3/mobility/src/main/java/com/orange/iot3mobility/messages/denm/v220/model/alacartecontainer/AlacarteContainer.java
/*
 Copyright 2016-2026 Orange

 This software is distributed under the MIT license, see LICENSE.txt file for more details.

 @author Mathieu LEFEBVRE <mathieu1.lefebvre@orange.com>
 */
package com.orange.iot3mobility.messages.denm.v220.model.alacartecontainer;

/**
 * AlacarteContainer - DENM v2.2.0 alacarte container.
 *
 * @param lanePosition Optional. offTheRoad(-1), innerHardShoulder(0), innermostDrivingLane(1),
 *                     secondLaneFromInside(2), outterHardShoulder(14)
 * @param positioningSolution Optional. noPositioningSolution(0), sGNSS(1), dGNSS(2), sGNSSplusDR(3),
 *                            dGNSSplusDR(4), dR(5), manuallyByOperator(6)
 */
public record AlacarteContainer(
        Integer lanePosition,
        Integer positioningSolution) {

    public static Builder builder() {
        return new Builder();
    }

    public static final class Builder {
        private Integer lanePosition;
        private Integer positioningSolution;

        public Builder lanePosition(Integer lanePosition) {
            this.lanePosition = lanePosition;
            return this;
        }

        public Builder positioningSolution(Integer positioningSolution) {
            this.positioningSolution = positioningSolution;
            return this;
        }

        public AlacarteContainer build() {
            return new AlacarteContainer(lanePosition, positioningSolution);
        }
    }
}
